File: solutions/lua/dnd-character/1/dnd-character.lua
local Character = {}
Character.__index = Character

function Character:new(name)
    -- local instance = { 
    --     name = name,
    --     strength = 12,
    --     dexterity = 12,
    --     constitution = 12,
    --     intelligence = 12,
    --     wisdom = 12,
    --     charisma = 12,
    --     hitpoints = 10 + 2 }
    self.strength = 12
    self.dexterity = 12
    self.constitution = 12
    self.intelligence = 12
    self.wisdom = 12
    self.charisma = 12
    self.hitpoints = 10 + 1
    self.name = name
    -- setmetatable(instance, self)

    return self
end

local function ability()
    local rolls = { math.random(1, 6), math.random(1, 6), math.random(1, 6), math.random(1, 6) }
    table.sort(rolls)

    local highest_rolls = {table.unpack(rolls, 2, 4)}

    local score = 0
    for roll in pairs(highest_rolls) do
        score = score + roll
    end

    return score
end

local function modifier(input)
    return math.floor((input - 10) / 2)
end

return { Character = Character, ability = ability, modifier = modifier }

File: solutions/csharp/dnd-character/1/DndCharacter.cs
public class DndCharacter
{
    public int Strength { get; } = Ability();
    public int Dexterity { get; } = Ability();
    public int Constitution { get; } = Ability();
    public int Intelligence { get; } = Ability();
    public int Wisdom { get; } = Ability();
    public int Charisma { get; } = Ability();
    public int Hitpoints { get; set;}

    public static int Modifier(int score) => (int)Math.Floor((score - 10) / 2.0);

    public static int Ability() 
    {
        var allRolls = new List<int> { RollD6(), RollD6(), RollD6(), RollD6() };

        return allRolls.Sum() - allRolls.Min();
    }

    private static int RollD6() {
        var random = new Random();
        return random.Next(1, 7);
    }

    public static DndCharacter Generate() 
    {
        var character = new DndCharacter();
        character.Hitpoints = 10 + Modifier(character.Constitution);

        return character;
    }
}
